website_sale_product_contract_gift/models/sale_order.py

```diff
@@ -0,0 +1,181 @@
+                        year=date_start.year + 1,
+                        month=date_start.month,
+                        day=date_start.day - 1,
+                    )
+                contract = contract_model.create(
+                    {
+                        "name": f"{line.product_id.name}: {order.name}",
+                        "partner_id": contract_partner.id,
+                        "contract_type": "sale",
+                        "is_gift": True,
+                        "date_start": order.gift_date,
+                        "payment_mode_id": order.payment_mode_id.id,
+                        "line_recurrence": True,
+                        "recurring_rule_type": "yearly",
+                        "recurring_interval": 1,
+                        "contract_line_ids": [
+                            fields.Command.create(
+                                {
+                                    "sale_order_line_id": line.id,
+                                    "product_id": line.product_id.id,
+                                    "name": line.product_id.name,
+                                    "quantity": line.product_uom_qty,
+                                    "uom_id": line.product_uom.id,
+                                    "price_unit": 0,
+                                    "date_start": date_start,
+                                    "date_end": date_end,
+                                    "recurring_rule_type": "yearly",
+                                    "recurring_interval": 1,
+                                    "auto_renew_rule_type": "yearly",
+                                    "auto_renew_interval": "1",
+                                    "recurring_next_date": line.date_start,
+                                }
+                            ),
+                        ],
+                    }
+                )
+                contract._onchange_contract_type()
+                # Cannot use "contract_id" field on order_line because
+                # for a gift the partner on sale.order is not the same
+                # as the partner on the contract.
+                line.write({"gift_contract_id": contract.id})
+            for line in line_to_update_contract:
+                line.create_contract_line(line.contract_id)
+        return super().action_confirm()
+
+    @api.model
+    def find_contact_partners(self, partner):
+        """Find partners that are the same as :partner:"""
+        partners_found = self.env["res.partner"].search(
+            [
+                ("email", "=", partner.email),
+                ("type", "=", "contact"),
+            ]
+        )
+        result = partners_found
+        fields_to_compare = [
+            "country_id",
+            "zip",
+            "name",
+            "user_id",
+        ]
+
+        if result:
+            for field in fields_to_compare:
+                filtered_partners = result.filtered(lambda r: r.mapped(field))
```

Comment:
how is this comparing the value of the field to the value of the field of the provided partner? shouldn’t this be something like this?
```suggestion
                filtered_partners = result.filtered(lambda r: r[field] == partner[field])
```

i’m not sure to understand how this whole function is supposed to work. a comment would help.

Reply:
Indeed, I added tests for this function, and explanation. Your suggestion is correct.